### ISSUE TITLE
Allow scope in BREAD relations

### ISSUE THREAD

**Issue by ghost** [thedevdojo/voyager#4627]
## Description of problem
I've got these relationships:

TextBlock  ----- belongsTo ----> Socket
Banner ----- belongsTo ----> Socket

Socket has an  `accepts` field, which prevents TextBlocks to be associated with some specific Sockets. For example, a socket "carousel" will not allow text, only banners.
Depending on the value of Socket::accepts, I want the socket to be shown or not in the list of Sockets, when creating/editing a text block.

## Proposed solution
Add an option `local scope` to the relation details.
I already have local scopes which filter sockets for text blocks and banners. If I can simply specify it to options, it could be passed to the model when displaying available sockets, preventing user tfrom making forbidden associations.

## Alternatives considered
I could create a custom formfield, obviously, but using a scope would make things much easier.

**Comment by MrCrayon** [comment]
I'll have to check how we could implement that. 

In the meanwhile did you try to extend relation method and apply a global scope there?

**Comment by ghost** [comment]
No, I did not... because I don't know what to edit or where :)
Which class holds that relation method?

**Comment by ghost** [comment]
Oh sorry...  The model's relation... Obviously... But I don't want to use a global scope...

**Comment by ghost** [comment]
Actually I'm not sure I understand what you mean... The scope need to be applied to Socket, while I'm creating, for instance, a TextBlock record... Or I'm missing your point.

**Comment by MrCrayon** [comment]
I'm not entirely sure how you want to use it, I thought you wanted to filter the options shown while you are creating/editing TextBlocks.

In that case that select is loaded by an ajax call managed by `VoyagerBaseController` method `relation`

**Comment by MrCrayon** [comment]
If that's the case you can create your custom controller as explained here:
https://voyager-docs.devdojo.com/customization/overriding-files

And override `relation` method with something like this:
```php
    public function relation(Request $request)
    {
        TextBlock::addGlobalScope(new SocketScope());

        return parent::relation($request);
    }
```

You might have to check if you can apply Scope based on call, or perform other checks, that's the base.

**Comment by MrCrayon** [comment]
Actually you probably want to apply a scope to Socket.
You need to adapt it based on your needs.

**Comment by ghost** [comment]
I give it a try and let you know. Anyway, having the scope in the options would be great ;)

**Comment by MrCrayon** [comment]
As a reminder that scope will be just a filter on visualization, if you absolutely need to be sure no wrong association is saved you'll also need to add a validation/check before saving.

**Comment by ghost** [comment]
Absolutely. Only admins (in my case only 2 persons, one of them being a basic website user) will edit BlockTexts, it's just to have the appropriate Sockets in the list, no hack is expected ;)

**Comment by ghost** [comment]
It works! Thanks a lot!

About the feature request, if you're interested, I can have a look at it and maybe provide a PR.
If there are things you think I should think about, please let me know. 

For now, I think there would modifications in:
- resources/views/formfields/relationship.blade.php
- resources/views/bread/edit-add.blade.php

Not sure there is anything in a controller, since the details are stored as JSON.

**Comment by MrCrayon** [comment]
> It works! Thanks a lot!

Good :+1: 

> For now, I think there would modifications in:
> 
> * ~resources/views/formfields/relationship.blade.php~
> * ~resources/views/bread/edit-add.blade.php~
 - Http/Controllers/VoyagerBaseController.php
 - documentation

I didn't look into it yet but my concern is we either implement it full circle:
 - filter on view
 - checks before saving

Or we need to make clear that it's only a cosmetic change and depending on use case checks before saving need to be implemented.

**Comment by ghost** [comment]
Checks before saving can be easily implemented with model::saving() method, triggered by Eloquent. It's still necessary to handle the exception raised if validation fails.

About `resources/views/bread/edit-add.blade.php`: isn't it the place a filed needs to be added to let user specify they want to apply a scope?

**Comment by MrCrayon** [comment]
> About `resources/views/bread/edit-add.blade.php`: isn't it the place a filed needs to be added to let user specify they want to apply a scope?

Nope, that would be `resources/views/tools/bread/edit-add.blade.php` or something like that inside tools

**Comment by MrCrayon** [comment]
Closed with #4851
